Update build envs to ubuntu 20.04. Change mingw build env to arch

jroweboy opened the pull request:
Thanks to @lat9nq for the base change to the mingw build env.

DarkLordZach commented on linux-fresh/Dockerfile:
```diff
@@ -1,5 +1,17 @@
-RUN apt-get update && apt-get -y full-upgrade
-RUN apt-get install --no-install-recommends -y build-essential libsdl2-dev libssl-dev python qtbase5-dev qtwebengine5-dev libqt5opengl5-dev wget git ccache cmake ninja-build
+RUN DEBIAN_FRONTEND=noninteractive apt-get update && apt-get -y full-upgrade
+RUN DEBIAN_FRONTEND=noninteractive apt-get install --no-install-recommends -y \
```
combine everything into one run command

jroweboy replied:
These are intentionally separate commands due to docker build caching. This lets us reuse the same build caches for all of the different linux images up to the point where the system is updated. https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#leverage-build-cache

DarkLordZach replied:
Considering that these images are probably built once every two months, it doesn't make sense to optimize for build speed. Instead, we should optimize for storage size as that can reduce build times on every build. 
Combining commands reduces layer count, reducing size and adding the rm /var/lib... also does that. Do note, that remove the apt lists or cleaning up pacman only has tangible benefits if its in the same layer, otherwise it just adds storage space.

jroweboy replied:
I was curious to see how much impact doing everything on a single line would bring, and it turns out it doesn't affect the final size of the image at all. :( Secretly I was hoping it would since the image is 900MB right now 
![image](https://user-images.githubusercontent.com/952515/80448978-26e55300-88db-11ea-9fbe-c85b7965eb8a.png)
(3 days ago is the old build and the single `RUN` line build is the newest one.) 

So its true that it reduces the number of intermediate docker images, but this only helps in size for the person running docker build, and doesn't affect the total size of the final image uploaded. Since it doesn't affect the final VM size to have them in separate `RUN` statements, I'm still leaning towards keeping them separate.

DarkLordZach commented on linux-clang-format/Dockerfile:
```diff
@@ -1,4 +1,4 @@
-FROM ubuntu:18.04
+FROM ubuntu:20.04
 MAINTAINER yuzu
 RUN useradd -m -s /bin/bash yuzu
```
combine everything into one run command

DarkLordZach commented on linux-mingw/Dockerfile:
```diff
@@ -1,12 +1,43 @@
-RUN apt-key adv --keyserver keyserver.ubuntu.com --recv '72931B477E22FEFD47F8DECE02FE5F12ADDE29B2' && apt-get update
-RUN apt-get install -y sdl2-mingw-w64 qt5base-mingw-w64 qt5tools-mingw-w64 libsamplerate-mingw-w64 qt5multimedia-mingw-w64
+# Add mingw-repo "ownstuff" is a AUR with an up to date mingw64
+RUN echo "[ownstuff]" >> /etc/pacman.conf \
```
combine everything into one run command

jroweboy replied:
I wrote a blurb about it elsewhere, but in Docker, its considered best practice to split them into different RUN statements, so even tho my last answer doesn't strictly apply here, its still best practices to keep them in individual RUN statements.

DarkLordZach commented on linux-mingw/Dockerfile:
```diff
@@ -1,12 +1,43 @@
-FROM ubuntu:18.04
+FROM archlinux:latest
```
use an explicit tag

jroweboy replied:
Arch linux is not like ubuntu, as you probably know, its a rolling release distribution, so it doesn't matter what tag you use, all the arch packages will end up the same after the pacman updates that are run below. So its better to just start from the latest tag anyway.

DarkLordZach commented on linux-mingw/Dockerfile:
```diff
@@ -1,12 +1,43 @@
+    && echo "SigLevel = Optional TrustAll" >> /etc/pacman.conf \
+    && echo "Server = https://martchus.no-ip.biz/repo/arch/ownstuff/os/\$arch" >> /etc/pacman.conf
+RUN pacman -Syu --noconfirm
+RUN pacman -Syu --noconfirm
```
is this duplicated on purpose?

jroweboy replied:
Yes, its a quirk of pacman where the first pacman call does a system upgrade and updates arch itself (after which you must kill your shell, which doesn't matter in Docker), and the second one will do a package upgrade and update all of the packages installed on the systeem.

lat9nq replied:
It's duplicated because sometimes pacman will only update key files the first time you update. Running it twice ensures that both keys and update files are acquired. On MSYS, you have to do it multiple times. In Linux, it seems to work if you do it just once, but I wanted to be sure.

lioncash replied:
Perhaps this should be documented in a comment above the two lines to make it obvious for those not acquainted with how pacman works.

DarkLordZach commented on linux-fresh/Dockerfile:
```diff
@@ -1,5 +1,17 @@
+    git \
+    ccache \
+    cmake \
+    ninja-build
```
add a `rm -rf /var/lib/apt/lists/*` to reduce final layer size

jroweboy replied:
As a not, the internet is kinda split on if you should remove these or not. The reason you might want to keep them on this container is for people that pull the container to build with it locally, then you'll want to keep the caches around. But I'm kinda leaning towards removing them since by and large, this is primarily used for CI, so lemme push that change real fast.

DarkLordZach commented:
I conducted a test to verify the claims of removing the apt lists and also combining into a single run command.
Here are the results:
```
linux-fresh                         both                a059598cc380        52 seconds ago      854MB
linux-fresh                         only_single_run     c64933a683a2        8 minutes ago       883MB
linux-fresh                         only_rm_lists       341ff250849f        16 minutes ago      883MB
linux-fresh                         master              736efbbcdad4        16 minutes ago      883MB
```
This was conducted on ubuntu server 18.04 using docker 18.09.7.
As the results show, neither of the two claims on their own have any benefit, but when combined result in a savings of ~30MB.

jroweboy commented:
Oh I finally see why you thought i wasn't already doing `rm lists` ... i sent the PR accidentally from my master branch instead of my branch that i'd been pushing updates to... :| 

Okay, so I see your results, but I haven't a clue what I did wrong, because I was already doing both in my test. Here was the commit that i added to remove any caches https://github.com/jroweboy/build-environments/commit/affcece6a6ffe9d98c399354d00bd26074d0995c and then I took this commit and combined all the RUN statements into one, and posted the results that I did (that there was no difference in final build size)

Heres is the file i was testing with:
```
FROM ubuntu:20.04
MAINTAINER yuzu
RUN useradd -m -s /bin/bash yuzu
RUN DEBIAN_FRONTEND=noninteractive apt-get update && apt-get -y full-upgrade && \
    DEBIAN_FRONTEND=noninteractive apt-get install --no-install-recommends -y \
    build-essential \
    libsdl2-dev \
    libssl-dev \
    python \
    qtbase5-dev \
    qtbase5-private-dev \
    qtwebengine5-dev \
    libqt5opengl5-dev \
    wget \
    git \
    ccache \
    cmake \
    ninja-build && \
    apt-get clean autoclean && apt-get autoremove --yes && rm -rf /var/lib/{apt,dpkg,cache,log}/
# cleanup all extra information that apt may be caching to save storage space
```
and this one doesn't have any difference than not combining all the RUN steps.

jroweboy commented:
Updated. Found that the reason rm wasn't removing anything was because shell expansion doesn't seem to work in the dockerfile. So I changed it to just delete each of them. The extra cleanup on linux-fresh saved an additional 9MB.
